### PR TITLE
report dependency for metadata

### PR DESCRIPTION
**Pull request by ansyral**
report the dependency from article to its nearest toc

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs [hunk 1]
```diff
@@ -219,6 +222,96 @@ private void RestructureItem(TocItemViewModel item, List<TocItemViewModel> items
...
+            models.RunAll(model =>
+            {
+                var wrapper = tocModelCache[model.OriginalFileAndType.FullPath];
+
+                // If the TOC file is referenced by other TOC, not report dependency
+                if (wrapper.IsReferenceToc)
+                {
+                    return;
+                }
+                var outputFile = GetOutputPath(model.FileAndType);
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs]: >outputFile [](start = 20, length = 10)

not used?

File: src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs [hunk 2]
```diff
@@ -219,6 +222,96 @@ private void RestructureItem(TocItemViewModel item, List<TocItemViewModel> items
             }
         }
 
+        private void ReportDependency(ImmutableList<FileModel> models, IHostService host, ImmutableDictionary<string, TocItemInfo> tocModelCache, int parallelism)
+        {
+            var nearest = new Dictionary<string, Toc>();
+            models.RunAll(model =>
+            {
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs]: indent

File: src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs [hunk 3]
```diff
@@ -219,6 +222,96 @@ private void RestructureItem(TocItemViewModel item, List<TocItemViewModel> items
...
+        private void UpdateNearestToc(IHostService host, TocItemViewModel item, FileModel toc, Dictionary<string, Toc> nearest)
+        {
+            var tocHref = item.TocHref;
+            var type = Utility.GetHrefType(tocHref);
+            if (type == HrefType.MarkdownTocFile || type == HrefType.YamlTocFile)
+            {
+                UpdateNearestTocCore(host, tocHref, toc, nearest);
+            }
+            else
+            {
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs]: else if

File: src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs [hunk 4]
```diff
@@ -219,6 +222,96 @@ private void RestructureItem(TocItemViewModel item, List<TocItemViewModel> items
...
+                var wrapper = tocModelCache[model.OriginalFileAndType.FullPath];
+
+                // If the TOC file is referenced by other TOC, not report dependency
+                if (wrapper.IsReferenceToc)
+                {
+                    return;
+                }
+                var outputFile = GetOutputPath(model.FileAndType);
+                var item = wrapper.Content;
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs]: what about not-in-toc articles?

File: src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs [hunk 5]
```diff
@@ -219,6 +222,96 @@ private void RestructureItem(TocItemViewModel item, List<TocItemViewModel> items
...
+            var tocOutputFile = GetOutputPath(toc.FileAndType);
+            FileAndType itemSource;
+            if (allSourceFiles.TryGetValue(item, out itemSource))
+            {
+                var itemOutputFile = GetOutputPath(itemSource);
+                var relative = tocOutputFile.RemoveWorkingFolder() - itemOutputFile;
+                Toc cur;
+                lock (nearest)
+                {
+                    if (!nearest.TryGetValue(item, out cur) || CompareRelativePath(relative, cur.OutputPath) < 0)
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs]: >CompareRelativePath [](start = 63, length = 19)

should report all the linked toc, not only the nearest?

Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs]: toc-> article this dependency is constructed later using LinkedToFiles and LinkedToUids

---
In reply to: [100465411](https://github.com/dotnet/docfx/pull/1285#discussion_r100465411) [](ancestors = 100465411)

File: src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs (hunk 4 above)
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs]: the only case that doesn't work for not-in-toc articles is that its nearest toc file is removed. Nice catch, i'll fix it

---
In reply to: [100463822](https://github.com/dotnet/docfx/pull/1285#discussion_r100463822) [](ancestors = 100463822)

File: src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs (hunk 5 above)
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocument.cs]: got it. TOC will always be built.

---
In reply to: [100466512](https://github.com/dotnet/docfx/pull/1285#discussion_r100466512) [](ancestors = 100466512,100465411)

File: src/Microsoft.DocAsCode.Build.Engine/CompilePhaseHandlerWithIncremental.cs [hunk 6]
```diff
@@ -240,6 +241,12 @@ private IEnumerable<DependencyItem> GetUidDependencyCore(FileModel model)
                         {
                             yield return new DependencyItem(sourceFile, f, sourceFile, DependencyTypeName.Uid);
                         }
+
+                        var name = Path.GetFileName(sourceFile);
+                        if (name.Equals("toc.md", StringComparison.OrdinalIgnoreCase) || name.Equals("toc.yml", StringComparison.OrdinalIgnoreCase))
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/CompilePhaseHandlerWithIncremental.cs]: add todo to remove this code.

**Comment by vwxyzh** [comment]
:shipit:

**Comment by superyyrrzz** [comment]
:shipit:

**Pull request merged**